pkg/kritis/signer/signer.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package signer
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/grafeas/kritis/pkg/kritis/apis/kritis/v1beta1"
+	"github.com/grafeas/kritis/pkg/kritis/crd/authority"
+	"github.com/grafeas/kritis/pkg/kritis/crd/vulnzsigningpolicy"
+	"github.com/grafeas/kritis/pkg/kritis/metadata"
+	"github.com/grafeas/kritis/pkg/kritis/secrets"
+	"github.com/grafeas/kritis/pkg/kritis/util"
+)
+
+type Signer struct {
+	config *Config
+	client metadata.ReadWriteClient
+}
+
+type Config struct {
+	Validate  vulnzsigningpolicy.ValidateFunc
+	PgpKey    *secrets.PgpKey
+	Authority v1beta1.AttestationAuthority
+	Project   string
+}
+
+func New(client metadata.ReadWriteClient, c *Config) Signer {
+	return Signer{
+		client: client,
+		config: c,
+	}
+}
+
+type ImageVulnerabilities struct {
+	ImageRef        string
+	Vulnerabilities []metadata.Vulnerability
+}
+
+// For testing
+var (
+	authFetcher = authority.Authority
+)
+
+// ValidateAndSign validates image from vulnz signing policy and then creates
+// attestation for the passing image.
+// Returns an error if image does not pass or creating an attestation fails.
+func (s Signer) ValidateAndSign(imageVulnz ImageVulnerabilities, vps v1beta1.VulnzSigningPolicy) error {
+	glog.Infof("Validating %q against VulnzSigningPolicy %q", imageVulnz.ImageRef, vps.Name)
+	violations, err := s.config.Validate(vps, imageVulnz.ImageRef, imageVulnz.Vulnerabilities)
+	if err != nil {
+		return fmt.Errorf("error when evaluating image %q against policy %q", imageVulnz.ImageRef, vps.Name)
+	}
+	if violations != nil {
+		return fmt.Errorf("image %q does not pass VulnzSigningPolicy %q: %v", imageVulnz.ImageRef, vps.Name, violations)
+	} else {
+		glog.Infof("Image %q passes VulnzSigningPolicy %s.", imageVulnz.ImageRef, vps.Name)
+		existed, _ := s.isAttestationAlreadyExist(imageVulnz.ImageRef)
```

Comment:
What if it exists, but is invalid (bad signature, whatever)?  Should we still return nil?

Reply:
Good question. 
The signer cannot verify whether the existing attestation is valid because it might not have the corresponding key. Because of this, replacing it with a new one can also be risky because the old one can be valid.
I think we should leave it as is, until we can provide more explicit control over the behavior to the users, e.g., a "--overwrite" flag. 
WDYT?

Reply:
Yup, I agree.